hyperstart/multicast.go

```
@@ -0,0 +1,158 @@
+			if err != nil {
+				glog.Errorf("Multicaster write error: %s\n", err)
+				break
+			}
+		}
+	}()
+
+	return ctlMulticast
+}
+
+func (m *multicast) buildEventID(containerID, processID string) string {
+	return fmt.Sprintf("%s-%s", containerID, processID)
+}
+
+func (m *multicast) sendEvent(msg *hyper.DecodedMessage) error {
+	var paeData hyper.ProcessAsyncEvent
+
+	err := json.Unmarshal(msg.Message, paeData)
+	if err != nil {
+		return err
+	}
+
+	uniqueID := m.buildEventID(paeData.Container, paeData.Process)
+	channel, exist := m.event[uniqueID]
+	if !exist {
+		return nil
+	}
+
+	channel <- msg
+
+	delete(m.event, uniqueID)
+
+	return nil
+}
+
+func (m *multicast) sendReply(msg *hyper.DecodedMessage) error {
+	var chanListLen int
+
+	m.Lock()
+	for {
+		chanListLen = len(m.reply)
+		if chanListLen > 0 {
+			break
+		}
+		m.Unlock()
+		time.Sleep(time.Millisecond)
+		m.Lock()
+	}
+
+	replyChannel := m.reply[0]
+
+	tmpChanList := make([]chan *hyper.DecodedMessage, chanListLen-1)
+	copy(tmpChanList, m.reply[1:])
+	m.reply = tmpChanList
+
+	m.Unlock()
+
+	// The current reply channel has been removed from the list, that's why
+	// we can be out of the mutex to send through that channel. Indeed, there
+	// is no risk that someone else tries to write on this channel.
+	replyChannel <- msg
+
+	return nil
+}
+
+func (m *multicast) write(msg *hyper.DecodedMessage) error {
+	switch msg.Code {
+	case hyper.INIT_NEXT:
+		return nil
+	case hyper.INIT_PROCESSASYNCEVENT:
+		return m.sendEvent(msg)
+	default:
+		return m.sendReply(msg)
+	}
+}
+
+func (m *multicast) listen(containerID, processID string, dataType ctlDataType) (chan *hyper.DecodedMessage, error) {
+	switch dataType {
+	case replyType:
+		newChan := make(chan *hyper.DecodedMessage)
+
+		m.Lock()
+		m.reply = append(m.reply, newChan)
+		m.Unlock()
+
+		return newChan, nil
+	case eventType:
+		uniqueID := m.buildEventID(containerID, processID)
+		m.event[uniqueID] = make(chan *hyper.DecodedMessage)
```

You want to check if `m.event[uniqueID]` is empty. If it's not, you should return an error or else you'll have a memory leak.